iroha/src/sumeragi.rs

```diff
@@ -31,59 +35,126 @@ trait Consensus {
+    /// This way peers vote for changing the view, if the leader refuses to accept this transaction.
+    pub fn sign(&mut self, public_key: &PublicKey, private_key: &PrivateKey) -> Result<(), String> {
+        let payload: Vec<u8> = self.transaction.clone().into();
+        if self.signatures.contains_key(public_key) {
+            Err("Already signed by these keys.".to_string())
+        } else {
+            self.signatures.insert(
+                *public_key,
+                Signature::new(*public_key, &payload, private_key)?,
+            );
+            Ok(())
+        }
+    }
+}
+
+/// Message structure describing a transaction that is forwarded from a client by a peer to the leader.
+#[derive(Io, Decode, Encode, Debug, Clone)]
+pub struct ForwardedTx {
+    transaction: AcceptedTransaction,
+    peer: PeerId,
+}
+
+/// Message structure describing a receipt sent by the leader to the peer it got this transaction from.
+#[derive(Io, Decode, Encode, Debug, Clone)]
+pub struct TxReceipt {
+    transaction_hash: Hash,
+    received_at: Duration,
+    signature: Signature,
+}
+
+impl TxReceipt {
+    /// Constructs a new receipt
+    pub fn new(
+        transaction: &AcceptedTransaction,
+        public_key: &PublicKey,
+        private_key: &PrivateKey,
+    ) -> Result<TxReceipt, String> {
+        let transaction_hash = transaction.hash();
+        Ok(TxReceipt {
+            transaction_hash,
+            received_at: SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("Failed to get System Time."),
+            signature: Signature::new(*public_key, &transaction_hash, private_key)?,
+        })
+    }
+}
+
 /// Message structure describing a request to other peers to change view because of the commit timeout.
 /// Peers vote on this view change by signing and forwarding this structure.
 #[derive(Io, Decode, Encode, Debug, Clone)]
 pub struct CommitTimeout {
     voting_block_hash: Hash,
-    signatures: Vec<Signature>,
+    signatures: BTreeMap<PublicKey, Signature>,
 }
 
 impl CommitTimeout {
     /// Constructs a new commit timeout message with no signatures.
     pub fn new(voting_block: VotingBlock) -> CommitTimeout {
         CommitTimeout {
             voting_block_hash: voting_block.block.hash(),
-            signatures: vec![],
+            signatures: BTreeMap::new(),
         }
     }
 
     /// Signes this request with the peer's public and private key.
     /// This way peers vote for changing the view.
     pub fn sign(&mut self, public_key: &PublicKey, private_key: &PrivateKey) -> Result<(), String> {
-        let already_signed = self
-            .signatures
-            .iter()
-            .any(|signature| signature.public_key == *public_key);
-        if !already_signed {
-            self.signatures.push(Signature::new(
+        if self.signatures.contains_key(public_key) {
+            Err("Already signed by these keys.".to_string())
```

Comment:
Do we need error in such a case? Maybe we can just leave as is or update signature?

Reply:
We don't need an explicit error, but we do need a result. We should know if we added a new signature or not, to decide whether to forward the message.